lightningnetwork/lnd
# routing: Support Pathfinding to Blinded Routes

## Change Description
This PR surfaces the ability to create routes to blinded path via the `QueryRoutes` api. It takes the approach of expressing a blinded path as a chain of hop hints, and backfills all the fields required for blinded payments in a second pass. 

Fixes #7200. 
Depends on [lightning-onion/57.](https://github.com/lightningnetwork/lightning-onion/pull/57)

## Steps to Test - LDK
Setup the following network + channels using my [branch of ldk-sample](https://github.com/carlaKC/ldk-sample/commit/94d1705e1f52bdc35cef42e605ca014b23280d5a): `LND --- LDK0 --- LDK1`

Create an invoice on LDK1: 
- `getinvoice (amt) (expiry seconds)` -> `lnr...`
- `lncli decodepayreq lnr...` -> get payment hash,  payment addr and cltv delta

Create a blinded path to LDK1 using tests: 
- Checkout [my branch of rust-lightning](https://github.com/carlaKC/rust-lightning/tree/route-blinding-interop) and open `lightning/src/blinded_path/test.rs`
- Update the following variables: 
  - `introduction_node` = LDK0 pubkey
  - `blinded_node` = LDK1 pubkey
  - `payment_secret` = payment addr from invoice
  - channel_id = channel connecting LDK0 --> LDK1
 - Run the test using `cargo test -- --nocapture test_blinded_path` -> this will print out a blinded path. 

Pay the blinded path via LND:
```
lncli queryroutes 
--introduction_point={cleartext_introduction_node_id} 
--blinding_point={blinding_point} 
--blinded_hops={blinded_introduction_node}:{introduction_node_data}
--blinded_hops={blinded_node}:{blinded_node_data}
--blinded_cltv= 120 + {invoice_cltv_delta}
--amt={invoice_amount} |
 lncli sendtoroute --payment_hash={payment_hash} -
```

The payment should succeed! 

Footguns: 
* CLTV:  make sure to add the delta in the LDK invoice to the value provided in the blinded path (which is just hardcoded to 120).
* Payment address: needs to be unique per-payment (for LDK reasons), so you have to regenerate every time.

## Steps to Test - CLN

<details>
  <summary>(this way is harder IMO, but was originally included in the PR description).</summary>
Since LND doesn't _produce_ blinded routes or forward blinded payments, you'll need the following setup: 
`Alice (LND) --- Bob(CLN) --- Carol (CLN) --privatechan-- Dave(CL)`

Run all CLN nodes in developer mode with: `--experimental-offers --dev-fast-gossip`.

You want an invoice with a blinded path, to do this you can create an offer and then fetch the raw invoice:
- `cl-dave offer 1000 test` -> `lno...` is your offer
- `cl-carol fetchinvoice lno...` -> `lni....` is your invoice
- `cl-carol decode lni...` -> This will give blinded hops/data etc. 

You can then pay the blinded invoice from LND: 
```
lncli queryroutes --introduction_point={cleartext_introduction_node_id} --blinding_point={blinding_point} --blinded_hops={blinded_introduction_node}:{introduction_node_data} --blinded_cltv={blinded_route_total_cltv} --amt={invoice_amount} | lncli sendtoroute --payment_hash={payment_hash} -
```

CLN only creates a blinded path when the node has private channels, and otherwise will just create an invoice with the receiving node as the introduction node. To test different variations of blinded paths: 
* With blinded hop: create offer on Dave and fetch with Carol
* With just introduction hop: create offer on Carol and fetch with Bob

</details>
